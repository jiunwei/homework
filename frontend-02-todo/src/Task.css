.Task {
	background: white;
  border-bottom: 1px solid silver;
  border-top: 1px solid silver;
	padding: 1.2em 1em;
}

.Task-Ghost {
  opacity: 0.5;
}
