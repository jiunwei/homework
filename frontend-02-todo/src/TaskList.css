.TaskList {
	flex: 1;
  margin: 0 0.5em;
	position: relative;
}

.TaskList > h1 {
  background: gray;
  color: white;
  margin: 0;
  font-size: 1.25em;
  padding: 1.3em 0.8em;
}

.TaskList-Counter {
	position: absolute;
	top: 0;
	right: 0;
}

.TaskList-Items {
  margin: 0;
  padding: 0;
}

.TaskList-Items > li {
  list-style: none;
  margin: 0;
  padding: 0;
}

.TaskList-Buffer {
  height: 4em;
}
