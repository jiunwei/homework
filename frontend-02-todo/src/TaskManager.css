.TaskManager {
	background: silver;
	padding: 1em;
	position: relative;
}

.TaskManager-EntryBar {
  float: left;
}

.TaskManager-Counter {
	float: right;
}


.TaskManager-Counter > h1 {
  font-size: 1em;
  margin: 0;
  margin-bottom: -0.5em;
  text-align: center;
  text-transform: uppercase;
}

.TaskManager-Items {
  clear: both;
	display: flex;
}
