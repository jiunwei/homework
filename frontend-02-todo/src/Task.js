import React, { Component } from 'react';
import { DragSource, DropTarget } from 'react-dnd';
import Constants from './Constants';
import './Task.css';

const sourceSpec = {
  beginDrag(props) {
    return {
      id: props.id,
      listIndex: props.listIndex,
      index: props.index
    }
  }
}

const targetSpec = {
  hover(props, monitor, component) {
    const dragItem = monitor.getItem();
    props.moveTask(dragItem.listIndex, dragItem.index, props.listIndex, props.index);
    monitor.getItem().listIndex = props.listIndex;
    monitor.getItem().index = props.index;
  }
}

class Task extends Component {
  render() {
  	const { id, title, connectSource, connectTarget, dragItem } = this.props;
    return (
      connectTarget(connectSource(<div className={(dragItem && dragItem.id === id) ? 'Task Task-Ghost' : 'Task'}>{title}</div>))
    );
  }
}

Task = DragSource(Constants.TASK, sourceSpec, (connect, monitor) => {
  return {
    connectSource: connect.dragSource()
  };
})(Task);

Task = DropTarget(Constants.TASK, targetSpec, (connect, monitor) => {
  return {
    connectTarget: connect.dropTarget(),
    dragItem: monitor.getItem()
  };
})(Task);

export default Task;
