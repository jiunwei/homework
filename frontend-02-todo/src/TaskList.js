import React, { Component } from 'react';
import { DropTarget } from 'react-dnd';
import Constants from './Constants';
import Counter from './Counter';
import Task from './Task';
import './TaskList.css';

const targetSpec = {
  hover(props, monitor, component) {
    const dragItem = monitor.getItem();
    // Don't count a task if we are appending to a list that it already belongs in.
    const targetIndex = (dragItem.listIndex === props.index) ? (props.tasks.length - 1) : props.tasks.length;
    props.moveTask(dragItem.listIndex, dragItem.index, props.index, targetIndex);
    monitor.getItem().listIndex = props.index;
    monitor.getItem().index = targetIndex;
  }
}

class TaskList extends Component {
  render() {
  	const { title, tasks, index, moveTask, connectTarget } = this.props;
  	const taskElements = tasks.map((task, i) => (<li key={i}><Task id={task.id} title={task.title} listIndex={index} index={i} moveTask={moveTask} /></li>));
    return (
    	<div className="TaskList">
        <h1>{title}</h1>
    		<div className="TaskList-Counter"><Counter count={tasks.length} unit="Project" /></div>
    		<ol className="TaskList-Items">
          {taskElements}
          {connectTarget(<li className="TaskList-Buffer">&nbsp;</li>)}
        </ol>
    	</div>
    );
  }
}

TaskList = DropTarget(Constants.TASK, targetSpec, (connect, monitor) => {
  return {
    connectTarget: connect.dropTarget()
  };
})(TaskList);

export default TaskList;
