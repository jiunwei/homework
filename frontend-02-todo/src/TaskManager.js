// Front-End Engineer Exercise: Todo
// Jiun Wei Chia
// 2017-03-23

import React, { Component, PropTypes } from 'react';
import { DragDropContext } from 'react-dnd';
import HTML5Backend from 'react-dnd-html5-backend';
import update from 'immutability-helper';
import EntryBar from './EntryBar';
import Counter from './Counter';
import TaskList from './TaskList';
import './TaskManager.css';

class TaskManager extends Component {
  static propTypes = {
    nextId: PropTypes.number,
    taskLists: PropTypes.arrayOf(PropTypes.object)
  };

  static defaultProps = {
    nextId: 0,
    taskLists: [
      {
        title: 'To do',
        tasks: []
      },
      {
        title: 'In Progress',
        tasks: []
      },
      {
        title: 'Done',
        tasks: []
      }
    ]
  };

  constructor(props) {
    super(props);
    this.state = {
      taskLists: this.props.taskLists
    }
    this.nextId = this.props.nextId;
    this.addNewTask = this.addNewTask.bind(this);
    this.moveTask = this.moveTask.bind(this);
  }

  addNewTask(title) {
    this.setState({
      taskLists: update(this.state.taskLists, {
        0: {
          tasks: {
            $push: [{ id: this.nextId, title: title }]
          }
        }
      })
    });
    ++this.nextId;
  }

  moveTask(fromListIndex, fromTaskIndex, toListIndex, toTaskIndex) {
    const movedTask = this.state.taskLists[fromListIndex].tasks[fromTaskIndex];
    if (fromListIndex === toListIndex) {
      if (fromTaskIndex === toTaskIndex) return;
      this.setState({
        taskLists: update(this.state.taskLists, {
          [fromListIndex]: {
            tasks: {
              $splice: [
                [fromTaskIndex, 1],
                [toTaskIndex, 0, movedTask]
              ]
            }
          }
        })
      });
    } else {
      this.setState({
        taskLists: update(this.state.taskLists, {
          [fromListIndex]: {
            tasks: {
              $splice: [[fromTaskIndex, 1]]
            }
          },
          [toListIndex]: {
            tasks: {
              $splice: [[toTaskIndex, 0, movedTask]]
            }
          }
        })
      });
    }
  }

  render() {
  	let count = 0;
  	const taskListElements = this.state.taskLists.map((taskList, i) => {
  		count += taskList.tasks.length;
  		return (<TaskList key={i} title={taskList.title} tasks={taskList.tasks} index={i} moveTask={this.moveTask} />);
  	});
    return (
    	<div className="TaskManager">
    		<div className="TaskManager-EntryBar">
	    	  <EntryBar prompt="add project" value="" onEnter={this.addNewTask} />
        </div>
	    	<div className="TaskManager-Counter">
	    		<h1>Total</h1>
	    		<Counter count={count} unit="Project" />
	    	</div>
	    	<div className="TaskManager-Items">
	    		{taskListElements}
	    	</div>
	    </div>
    );
  }
}

export default DragDropContext(HTML5Backend)(TaskManager);
