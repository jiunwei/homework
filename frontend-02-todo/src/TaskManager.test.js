// Front-End Engineer Exercise: Todo Unit Tests
// Jiun Wei Chia
// 2017-03-23

import React from 'react';
import ReactTestUtils from 'react-addons-test-utils'
import renderer from 'react-test-renderer';
import { DragDropContext } from 'react-dnd';
import TestBackend from 'react-dnd-test-backend';
import TaskManager from './TaskManager';
import TaskList from './TaskList';
import Task from './Task';

const SAMPLE_TASKS = [
  {
    title: 'To do',
    tasks: [
      { id: 0, title: 'Project 1' },
      { id: 1, title: 'Project 2' },
      { id: 2, title: 'Project 3' },
      { id: 3, title: 'Project 4' }
    ]
  },
  {
    title: 'In Progress',
    tasks: [
      { id: 4, title: 'Project 5' },
      { id: 5, title: 'Project 6' },
      { id: 6, title: 'Project 7' }
    ]
  },
  {
    title: 'Done',
    tasks: [
      { id: 7, title: 'Project 8' },
      { id: 8, title: 'Project 9' },
      { id: 9, title: 'Project 10' },
      { id: 10, title: 'Project 11' },
      { id: 11, title: 'Project 12' }
    ]
  }
];

describe('TaskManager', () => {

  it('renders a set of empty tasks correctly', () => {
    const layout = renderer.create(<TaskManager />).toJSON();
    expect(layout).toMatchSnapshot();
  });

  it('renders a set of sample tasks correctly', () => {
    const layout = renderer.create(<TaskManager nextId={12} taskLists={SAMPLE_TASKS} />).toJSON();
    expect(layout).toMatchSnapshot();
  });

  describe('#addNewTask()', () => {

    it('adds tasks to the first task list when called directly', () => {
      // Need to unwrap drag-and-drop context using getDecoratedComponentInstance().
      const taskManager = ReactTestUtils.renderIntoDocument(<TaskManager nextId={12} taskLists={SAMPLE_TASKS} />).getDecoratedComponentInstance();
      const beforeCounts = taskManager.state.taskLists.map((taskList) => taskList.tasks.length);

      taskManager.addNewTask('Project 1');
      taskManager.addNewTask('Project 2');
      taskManager.addNewTask('Project 3');

      const taskLists = taskManager.state.taskLists;
      const afterCounts = taskLists.map((taskList) => taskList.tasks.length);
      expect(afterCounts[0]).toEqual(beforeCounts[0] + 3);
      expect(afterCounts[1]).toEqual(beforeCounts[1]);
      expect(afterCounts[2]).toEqual(beforeCounts[2]);
      const lastTasks = taskLists[0].tasks.slice(-3);
      expect(lastTasks[0].title).toEqual('Project 1');
      expect(lastTasks[1].title).toEqual('Project 2');
      expect(lastTasks[2].title).toEqual('Project 3');
    });

    it('adds tasks to the first task list when called via the EntryBar', () => {
      const taskManager = ReactTestUtils.renderIntoDocument(<TaskManager nextId={12} taskLists={SAMPLE_TASKS} />);
      const taskLists = ReactTestUtils.scryRenderedComponentsWithType(taskManager, TaskList);
      const beforeCounts = taskLists.map((component) => component.props.tasks.length);

      const form = ReactTestUtils.findRenderedDOMComponentWithTag(taskManager, 'form');
      const input = ReactTestUtils.findRenderedDOMComponentWithTag(taskManager, 'input');
      ReactTestUtils.Simulate.change(input, {target: {value: 'Project 1'}});
      ReactTestUtils.Simulate.submit(form);
      ReactTestUtils.Simulate.change(input, {target: {value: 'Project 2'}});
      ReactTestUtils.Simulate.submit(form);
      ReactTestUtils.Simulate.change(input, {target: {value: 'Project 3'}});
      ReactTestUtils.Simulate.submit(form);

      const afterCounts = taskLists.map((component) => component.props.tasks.length);
      expect(afterCounts[0]).toEqual(beforeCounts[0] + 3);
      expect(afterCounts[1]).toEqual(beforeCounts[1]);
      expect(afterCounts[2]).toEqual(beforeCounts[2]);
      const lastTasks = taskLists[0].props.tasks.slice(-3);
      expect(lastTasks[0].title).toEqual('Project 1');
      expect(lastTasks[1].title).toEqual('Project 2');
      expect(lastTasks[2].title).toEqual('Project 3');
    });

  });

  describe('#moveTask()', () => {

    it('moves a task to the right location when called directly', () => {
      // Need to unwrap drag-and-drop context using getDecoratedComponentInstance().
      const taskManager = ReactTestUtils.renderIntoDocument(<TaskManager nextId={12} taskLists={SAMPLE_TASKS} />).getDecoratedComponentInstance();
      const beforeCounts = taskManager.state.taskLists.map((taskList) => taskList.tasks.length);

      taskManager.moveTask(0, 1, 2, 2);
      taskManager.moveTask(1, 1, 1, 2);
      taskManager.moveTask(1, 2, 1, 2);
      taskManager.moveTask(2, 1, 1, 0);

      const taskLists = taskManager.state.taskLists;
      const afterCounts = taskLists.map((taskList) => taskList.tasks.length);
      expect(afterCounts[0]).toEqual(beforeCounts[0] - 1);
      expect(afterCounts[1]).toEqual(beforeCounts[1] + 1);
      expect(afterCounts[2]).toEqual(beforeCounts[2]);
      expect(taskLists[0].tasks[0].title).toEqual('Project 1');
      expect(taskLists[0].tasks[1].title).toEqual('Project 3');
      expect(taskLists[0].tasks[2].title).toEqual('Project 4');
      expect(taskLists[1].tasks[0].title).toEqual('Project 9');
      expect(taskLists[1].tasks[1].title).toEqual('Project 5');
      expect(taskLists[1].tasks[2].title).toEqual('Project 7');
      expect(taskLists[1].tasks[3].title).toEqual('Project 6');
      expect(taskLists[2].tasks[0].title).toEqual('Project 8');
      expect(taskLists[2].tasks[1].title).toEqual('Project 2');
      expect(taskLists[2].tasks[2].title).toEqual('Project 10');
      expect(taskLists[2].tasks[3].title).toEqual('Project 11');
      expect(taskLists[2].tasks[4].title).toEqual('Project 12');
    });

    it('moves a task to the right location when called via drag and drop', () => {
      const TestTaskManager = DragDropContext(TestBackend)(
        class TestTaskManager extends React.Component {
          render() {
            return (<TaskManager.DecoratedComponent nextId={12} taskLists={SAMPLE_TASKS} />);
          }
        }
      );
      const taskManager = ReactTestUtils.renderIntoDocument(<TestTaskManager />);
      const backend = taskManager.getManager().getBackend();
      const taskLists = ReactTestUtils.scryRenderedComponentsWithType(taskManager, TaskList);
      const beforeCounts = taskLists.map((component) => component.props.tasks.length);
      const beforeTasks = taskLists.map((taskList) => ReactTestUtils.scryRenderedComponentsWithType(taskList, Task));

      // Simulate first drag-and-drop in more detail.
      const source = beforeTasks[0][1];
      const intervening = beforeTasks[1][2];
      const target = beforeTasks[2][2];
      // Need to unwrap the DragTarget to reach the DragSource handler ID.
      backend.simulateBeginDrag([source.getDecoratedComponentInstance().getHandlerId()]);
      backend.simulateHover([source.getHandlerId()]);
      backend.simulateHover([intervening.getHandlerId()]);
      backend.simulateHover([target.getHandlerId()]);
      backend.simulateDrop();
      backend.simulateEndDrag();

      // Simulate remaining drag-and-drop actions.
      backend.simulateBeginDrag([beforeTasks[1][1].getDecoratedComponentInstance().getHandlerId()]);
      backend.simulateHover([beforeTasks[1][2].getHandlerId()]);
      backend.simulateDrop();
      backend.simulateEndDrag();
      backend.simulateBeginDrag([beforeTasks[1][2].getDecoratedComponentInstance().getHandlerId()]);
      backend.simulateHover([beforeTasks[1][2].getHandlerId()]);
      backend.simulateDrop();
      backend.simulateEndDrag();
      backend.simulateBeginDrag([beforeTasks[2][1].getDecoratedComponentInstance().getHandlerId()]);
      backend.simulateHover([beforeTasks[1][0].getHandlerId()]);
      backend.simulateDrop();
      backend.simulateEndDrag();

      const afterTasks = taskLists.map((taskList) => ReactTestUtils.scryRenderedComponentsWithType(taskList, Task));
      const afterCounts = taskLists.map((component) => component.props.tasks.length);
      expect(afterCounts[0]).toEqual(beforeCounts[0] - 1);
      expect(afterCounts[1]).toEqual(beforeCounts[1] + 1);
      expect(afterCounts[2]).toEqual(beforeCounts[2]);
      expect(afterTasks[0][0].props.title).toEqual('Project 1');
      expect(afterTasks[0][1].props.title).toEqual('Project 3');
      expect(afterTasks[0][2].props.title).toEqual('Project 4');
      expect(afterTasks[1][0].props.title).toEqual('Project 9');
      expect(afterTasks[1][1].props.title).toEqual('Project 5');
      expect(afterTasks[1][2].props.title).toEqual('Project 7');
      expect(afterTasks[1][3].props.title).toEqual('Project 6');
      expect(afterTasks[2][0].props.title).toEqual('Project 8');
      expect(afterTasks[2][1].props.title).toEqual('Project 2');
      expect(afterTasks[2][2].props.title).toEqual('Project 10');
      expect(afterTasks[2][3].props.title).toEqual('Project 11');
      expect(afterTasks[2][4].props.title).toEqual('Project 12');
    });

  });

});
