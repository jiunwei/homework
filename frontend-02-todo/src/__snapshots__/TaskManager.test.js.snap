exports[`TaskManager renders a set of empty tasks correctly 1`] = `
<div
  className="TaskManager">
  <div
    className="TaskManager-EntryBar">
    <form
      className="EntryBar"
      onSubmit={[Function]}>
      <span
        className="EntryBar-Prompt">
        add project
      </span>
      <input
        onChange={[Function]}
        type="text"
        value="" />
    </form>
  </div>
  <div
    className="TaskManager-Counter">
    <h1>
      Total
    </h1>
    <div
      className="Counter">
      <p
        className="Counter-Count">
        0
      </p>
      <p
        className="Counter-Unit">
        Projects
      </p>
    </div>
  </div>
  <div
    className="TaskManager-Items">
    <div
      className="TaskList">
      <h1>
        To do
      </h1>
      <div
        className="TaskList-Counter">
        <div
          className="Counter">
          <p
            className="Counter-Count">
            0
          </p>
          <p
            className="Counter-Unit">
            Projects
          </p>
        </div>
      </div>
      <ol
        className="TaskList-Items">
        <li
          className="TaskList-Buffer">
           
        </li>
      </ol>
    </div>
    <div
      className="TaskList">
      <h1>
        In Progress
      </h1>
      <div
        className="TaskList-Counter">
        <div
          className="Counter">
          <p
            className="Counter-Count">
            0
          </p>
          <p
            className="Counter-Unit">
            Projects
          </p>
        </div>
      </div>
      <ol
        className="TaskList-Items">
        <li
          className="TaskList-Buffer">
           
        </li>
      </ol>
    </div>
    <div
      className="TaskList">
      <h1>
        Done
      </h1>
      <div
        className="TaskList-Counter">
        <div
          className="Counter">
          <p
            className="Counter-Count">
            0
          </p>
          <p
            className="Counter-Unit">
            Projects
          </p>
        </div>
      </div>
      <ol
        className="TaskList-Items">
        <li
          className="TaskList-Buffer">
           
        </li>
      </ol>
    </div>
  </div>
</div>
`;

exports[`TaskManager renders a set of sample tasks correctly 1`] = `
<div
  className="TaskManager">
  <div
    className="TaskManager-EntryBar">
    <form
      className="EntryBar"
      onSubmit={[Function]}>
      <span
        className="EntryBar-Prompt">
        add project
      </span>
      <input
        onChange={[Function]}
        type="text"
        value="" />
    </form>
  </div>
  <div
    className="TaskManager-Counter">
    <h1>
      Total
    </h1>
    <div
      className="Counter">
      <p
        className="Counter-Count">
        12
      </p>
      <p
        className="Counter-Unit">
        Projects
      </p>
    </div>
  </div>
  <div
    className="TaskManager-Items">
    <div
      className="TaskList">
      <h1>
        To do
      </h1>
      <div
        className="TaskList-Counter">
        <div
          className="Counter">
          <p
            className="Counter-Count">
            4
          </p>
          <p
            className="Counter-Unit">
            Projects
          </p>
        </div>
      </div>
      <ol
        className="TaskList-Items">
        <li>
          <div
            className="Task">
            Project 1
          </div>
        </li>
        <li>
          <div
            className="Task">
            Project 2
          </div>
        </li>
        <li>
          <div
            className="Task">
            Project 3
          </div>
        </li>
        <li>
          <div
            className="Task">
            Project 4
          </div>
        </li>
        <li
          className="TaskList-Buffer">
           
        </li>
      </ol>
    </div>
    <div
      className="TaskList">
      <h1>
        In Progress
      </h1>
      <div
        className="TaskList-Counter">
        <div
          className="Counter">
          <p
            className="Counter-Count">
            3
          </p>
          <p
            className="Counter-Unit">
            Projects
          </p>
        </div>
      </div>
      <ol
        className="TaskList-Items">
        <li>
          <div
            className="Task">
            Project 5
          </div>
        </li>
        <li>
          <div
            className="Task">
            Project 6
          </div>
        </li>
        <li>
          <div
            className="Task">
            Project 7
          </div>
        </li>
        <li
          className="TaskList-Buffer">
           
        </li>
      </ol>
    </div>
    <div
      className="TaskList">
      <h1>
        Done
      </h1>
      <div
        className="TaskList-Counter">
        <div
          className="Counter">
          <p
            className="Counter-Count">
            5
          </p>
          <p
            className="Counter-Unit">
            Projects
          </p>
        </div>
      </div>
      <ol
        className="TaskList-Items">
        <li>
          <div
            className="Task">
            Project 8
          </div>
        </li>
        <li>
          <div
            className="Task">
            Project 9
          </div>
        </li>
        <li>
          <div
            className="Task">
            Project 10
          </div>
        </li>
        <li>
          <div
            className="Task">
            Project 11
          </div>
        </li>
        <li>
          <div
            className="Task">
            Project 12
          </div>
        </li>
        <li
          className="TaskList-Buffer">
           
        </li>
      </ol>
    </div>
  </div>
</div>
`;
